src/main/java/com/erigitic/jobs/TEAction.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * This file is part of Total Economy, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) Eric Grandt <https://www.ericgrandt.com>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.erigitic.jobs;
+
+import com.google.inject.Inject;
+import ninja.leaping.configurate.ConfigurationNode;
+import org.slf4j.Logger;
+import org.spongepowered.api.block.BlockState;
+import org.spongepowered.api.block.trait.BlockTrait;
+import org.spongepowered.api.data.DataQuery;
+
+import java.util.*;
+
+public class TEAction {
+
+    public static TEAction fromActionNode(String action, ConfigurationNode node) {
+        ConfigurationNode growthTraitNode = node.getNode("growthTrait");
+        if (growthTraitNode == null)
+            growthTraitNode = node.getNode("growth-trait");
+        ConfigurationNode idTraitNode = node.getNode("id-trait");
+        if (idTraitNode.isVirtual()) {
+            return new TEAction(action, node.getKey().toString(), TEActionReward.getFromNode(node), growthTraitNode.getString(null));
+        }
+        String traitName = idTraitNode.getString("type");
+        if (traitName == null)
+            return new TEAction();
+        ConfigurationNode idTraits = node.getNode(traitName);
+        Map<String, TEActionReward> rewardMap = new HashMap<>();
+        if (idTraits.hasMapChildren()) {
+            idTraits.getChildrenMap().forEach((k, v) -> {
+                if (!(k instanceof String))
+                    return;
+                TEActionReward reward = TEActionReward.getFromNode(v);
+                rewardMap.put(((String) k), reward);
+            });
+        }
+        return new TEAction(action, node.getKey().toString(), rewardMap, idTraitNode.getString("type"), growthTraitNode.getString(null));
+    }
+
+    private String action;
```

Comment:
Instance variables should be at the top of the class

Reply:
Ahead of constructors but after the statics - those are gone now though